Trustcoin tests

morelazers opened the pull request:
Includes some of @shine2lay and @rmerom's changes to make this Truffle 2 compatible.

I still need to add way more edgecases, and unfortunately I had to add a constructor on the new token contract, I will search for ways to get rid of this, and add a bunch more tests when I can.

I tried to split them into files detailing the exact parts of the code to test. There's way too much duplicated code right now, so I'll work on refactoring that at some point too.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/wetrustplatform/token-contracts/8)
<!-- Reviewable:end -->

morelazers committed:
Initial tests

morelazers committed:
Fix the incorrect send test

morelazers committed:
Merge branch 'develop' into trustcoin_tests

morelazers committed:
Merge branch 'develop' into trustcoin_tests

morelazers committed:
Fix some tests - and add some tests for the auth modifiers

morelazers committed:
Be a little more stringent with assertions

morelazers committed:
Fixes all for Truffle 2

rmerom committed:
cleanup and add some tools and other stuff

morelazers committed:
They pass at last! Should probably be modified so that the new contract doesn't need the contructor param, but for testing it's necessary right now.

morelazers committed:
Add tests for migrate functionality

morelazers committed:
Delete /app folder

rmerom commented:
Reviewed 2 of 20 files at r1, 1 of 9 files at r2, 1 of 9 files at r3.
Review status: 4 of 20 files reviewed at latest revision, 23 unresolved discussions.

---

*[contracts/ExampleTrustcoin2.sol, line 39 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLNxDimB867JxPGn6:-KfJLNxDimB867JxPGn7:b-3uuzkq) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/ExampleTrustcoin2.sol#L39)):*
> ```
>   }
> 
>   function ExampleTrustcoin2(address _oldTokenAddress) {
> ```

i'm not sure i understand what's wrong with having a constructor that takes the `_oldTokenAddress` , as long as `oldTokenAddress` is public and accessible?

---

*[contracts/Trustcoin.sol, line 26 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLZBsD46zTPex2Evo:-KfJLZBsD46zTPex2Evp:b-woau1b) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L26)):*
> ```
>   uint256 public totalSupply = 100000000 * (10 ** decimals); // One hundred million (ERC20)
>   uint256 public totalMigrated; // Begins at 0 and increments as tokens are migrated to a new contract
>   address public newTokenAddress;
> ```

didn't we agree this can be removed?

---

*[contracts/Trustcoin.sol, line 98 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLbOlj5H39Znf6zYP:-KfJLbOlj5H39Znf6zYQ:b-75ywm7) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L98)):*
> ```
>   }
> 
>   function blockTime() external returns (uint256) {
> ```

why is this needed?

---

*[contracts/Trustcoin.sol, line 112 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLguP4Y1gauWIhKbS:-KfJLguP4Y1gauWIhKbT:b-825yla) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L112)):*
> ```
>   // See OutgoingMigrationTokenInterface
>   function beginMigrationPeriod(address _newTokenAddress) onlyFromMigrationMaster external {
>     if (allowOutgoingMigrations) throw; // Ensure we haven't already started allowing migrations
> ```

i think we should also assert that `newTokenAddress == 0`. Otherwise post-migration we can re-start migration.
If that's true, can you add that to the list of edgecase tests?

---

*[contracts/deps/OutgoingMigrationTokenInterface.sol, line 15 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJM0cDtz3JGJe9iThI:-KfJM0cDtz3JGJe9iThJ:b-vz4kbj) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/deps/OutgoingMigrationTokenInterface.sol#L15)):*
> ```
>   /// Getter function for the new token address
>   /// @return address Address of the new token
>   address public newTokenAddress;
> ```

these comments need refreshing... and this var isn't really used by the derived contract, so maybe we should rethink about how to do this?

---

*[test/checkAuth.js, line 1 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJzsmkjAr_96rwpq7k:-KfJzsmlH8cLHPejv3ff:b-8o6uzf) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L1)):*
> ```JavaScript
> 'use strict'
> ```

nit: since all tests start with `check`, can we drop it? all of them sit in the`test` directory so we know what they are about.
nit2: I would rather "auth" be expanded to "authentication" as i wasn't sure if it's "authentication" or "authorization" until reading the code (really nitty nit)

---

*[test/checkAuth.js, line 18 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK-JgPmTpJnqGdf_Jh:-KfK-JgQnAz4o1XpTqCw:b-fbcyev) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L18)):*
> ```JavaScript
>     assert.equal(contractMigrationMaster, migrationMaster)
>   }))
>   it("should allow migration master to be changed by migration master", co(function* () {
> ```

nit: would be nice to have 1 blank line between tests (here and in other test files), would also be consistent with rosca-contracts

---

*[test/checkAuth.js, line 40 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK-lsmm_0Asc_QbmQf:-KfK-lsqq2m1CbS3e3rL:bv317cn) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L40)):*
> ```JavaScript
>     yield trst.beginMigrationPeriod(consts.NON_ZERO_ADDRESS, {from: migrationMaster})
>     let allowMigrations = yield trst.allowOutgoingMigrations.call()
>     assert.equal(allowMigrations, true)
> ```

nit: can use here and elsewhere `assert.isOk` for equal true, and `isNotOk` for equal false

---

*[test/checkAuth.js, line 47 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK0DkU17vqZwdYB8xg:-KfK0DkU17vqZwdYB8xh:bjqeztc) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L47)):*
> ```JavaScript
>     let trst = yield utils.deployTrustcoin(owner, migrationMaster)
>     yield trst.beginMigrationPeriod(consts.NON_ZERO_ADDRESS, {from: migrationMaster})
>     utils.increaseTime(consts.ONE_YEAR_IN_SECONDS)
> ```

nit: you might want to query for `trst.minimumMigrationDuration` in a `before()` method and put it into a var `x`. Then use `x + <1 week>` when chcecking these things. Otherwise we have a constant here that really depends on a constant in the contract and may break if we decide to change the contract.

---

*[test/checkInitialBalance.js, line 9 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK0lUcotwW65WC-IIe:-KfK0lUcotwW65WC-IIf:b-k076vt) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L9)):*
> ```JavaScript
> let consts = require("./utils/consts.js")
> 
> const TOTAL_SUPPLY = 100000000000000 // 1e14
> ```

same here as above, better use `trst.totalSupply.call()`. Future cloners of our contract will thank you (even if you won't know that), and also future readers of this test (who won't need to guess where this number came from).

---

*[test/checkInitialBalance.js, line 14 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK11NUWPW6al99zyT4:-KfK11NUWPW6al99zyT5:bpivep1) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L14)):*
> ```JavaScript
> 
>   // accounts_ is all the accounts that we have in testrpc
>   it("should put " + TOTAL_SUPPLY + " (one hundred million plus 6 decimals) Trustcoin in the first account", co(function* () {
> ```

"in the first account" - does that mean `msg.sender` ?

---

*[test/checkInitialBalance.js, line 17 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK19FhYU3iS-6Po_DV:-KfK19FhYU3iS-6Po_DW:b-n34375) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L17)):*
> ```JavaScript
>     let trst = yield utils.deployTrustcoin(accounts_[0], accounts_[1])
>     let balance = yield trst.balanceOf.call(accounts_[0])
>     return assert.equal(balance.valueOf(), TOTAL_SUPPLY, TOTAL_SUPPLY + " wasn't in the first account")
> ```

I find that strings (here and below) that explain trivial things in the code just add more clutter. if a test fails, dev will go and find the line anyway (because they'll need to fix) and will likely understand from the code w/o needing the explanation.
I would start with having no string at all, and explain only when the test code is complicated enough for a reader not to understand what's going on.

---

*[test/checkInitialBalance.js, line 27 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK1Pl-yf5GIPIahVYy:-KfK1Pl-yf5GIPIahVYz:b-p17ahr) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L27)):*
> ```JavaScript
>     let version = yield trst.version.call()
> 
>     assert.equal(name, 'Trustcoin', "Trustcoin wasn't the found name")
> ```

sorry, today I'm writing all these little things that you need to change if you don't want to :)

this is testing trivial code. If production code says `const a=6`, there's no point usually in testing that `a == 6`. It does add clutter, make maintainability harder (need to change in 2 places), and does not really check that a sophisticated piece of code is doing what it should. 
I think many of the following tests (though not all) might fall under this criterion.

---

*[test/checkInitialBalance.js, line 39 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK27gWHpMGVYd-WpWf:-KfK27gWHpMGVYd-WpWg:b-jh99ow) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L39)):*
> ```JavaScript
>   }))
> 
>   it("should not allow outgoing migrations", co(function* () {
> ```

... upon initialization?

---

*[test/checkMigrateTokens.js, line 37 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK37rnU176VujFFp5Y:-KfK37rnU176VujFFp5Z:b-ql49ee) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkMigrateTokens.js#L37)):*
> ```JavaScript
>     let trst2 = yield utils.deployExampleTrustcoin2(owner, trst.address)
>     yield trst.beginMigrationPeriod(trst2.address, {from: migrationMaster})
>     yield utils.assertThrows(trst.migrateToNewContract(tokensToMigrate, {from: migrationMaster}))
> ```

this is very specific. IIUC 1) the migrator is the migrationMaster (can we do someone else?) and 2) their balance is zero (can we have balance `x != 0` and try to migrate `x+1`)?

---

*[test/checkMigrateTokens.js, line 48 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK3ZCu0pw-yteUtqMU:-KfK3ZCv0-XB5pi0lvbY:bh6ckko) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkMigrateTokens.js#L48)):*
> ```JavaScript
>     let trst2 = yield utils.deployExampleTrustcoin2(owner, trst.address)
>     yield trst.beginMigrationPeriod(trst2.address, {from: migrationMaster})
>     yield trst.migrateToNewContract(tokensToMigrate, {from: owner})
> ```

also pretty specific. can we transfer `x` to some account other than `owner`, then migrate `x` to new coin?

---

*[test/checkMigrateTokens.js, line 66 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK3rKSzvLMwoUHx_B4:-KfK3rKSzvLMwoUHx_B5:b1ozndx) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkMigrateTokens.js#L66)):*
> ```JavaScript
>     assert.equal(newTokenSupply, 0)
>   }))
> })
> ```

I would add one longer test of a "normal scenario"
- migration begins
- people still transfer money in the old contract
- some people migrate, check `totalMigrated`and `totalSupply` on both
- migration is finalized
- no one can migrate, but can still `transfer` in both old and new contracts

perhaps in its own e2e test file?

---

*[test/checkMigrationPeriod.js, line 15 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4AC-9UqZafzmZLVo:-KfK4AC-9UqZafzmZLVp:bl789ip) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkMigrationPeriod.js#L15)):*
> ```JavaScript
>     let trst = yield utils.deployTrustcoin(owner, migrationMaster)
>     let trst2 = yield utils.deployExampleTrustcoin2(owner, trst.address)
>     yield trst.beginMigrationPeriod(trst2.address, {from: migrationMaster})
> ```

this can be made tighter - that is, wait `minMigrationPeriod - safetyDelta` and see that we still can't be migrate

---

*[test/checkSend.js, line 14 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4SjzbOZzRi2RpILs:-KfK4Sk-SpsurFWVt6i-:b3uiy1j) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkSend.js#L14)):*
> ```JavaScript
>     let amount = 10
>     let trst = yield utils.deployTrustcoin(accounts_[0], accounts_[1])
>     let accOneStartingBal = yield trst.balanceOf.call(accOne)
> ```

many style guides discourage shortening words (e.g. "acc" for "account") because reader has to guess what the writer meant. If you don't want to make the var name too long, consider `account1BalanceStart`. same for other `acc` vars

---

*[test/checkSend.js, line 20 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4uvn7kUwhhyIO5gO:-KfK4uvoidhcmLS0Ycvy:b-wuocr5) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkSend.js#L20)):*
> ```JavaScript
>     let accTwoEndingBal = yield trst.balanceOf.call(accTwo)
>     assert.equal(accOneEndingBal.toNumber(), accOneStartingBal.toNumber() - amount, "Amount wasn't correctly taken from the sender")
>     assert.equal(accTwoEndingBal.toNumber(), accTwoStartingBal.toNumber() + amount, "Amount wasn't correctly sent to the receiver")
> ```

silly, but maybe also check `totalSupply` hasn't changed?

---

*[test/utils/utils.js, line 10 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJzOi893xAVt3taYvs:-KfJzOi893xAVt3taYvt:b-1vz94) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/utils/utils.js#L10)):*
> ```JavaScript
> let Promise = require("bluebird")
> 
> // we need this becaues test env is different than script env
> ```

s/becaues/because

(also I'm not sure i understand what this means)

---

*[test/utils/utils.js, line 34 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJz_0AEd4aqrz80Ysa:-KfJz_0AEd4aqrz80Ysb:b-3qxcxr) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/utils/utils.js#L34)):*
> ```JavaScript
>       assert.isNotOk(true, err)
>     }).catch(function(e) {
>       assert.notInclude(e.message, 'invalid JUMP', "Invalid Jump error occurred")
> ```

unless we expect some exception to happen, can we rehtrow here?
otherwise we're swallowing exceptions

---

*[tools/build/run-tests.sh, line 3 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK501RuwLXBpklm4Ld:-KfK501RuwLXBpklm4Le:bfrtw23) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/tools/build/run-tests.sh#L3)):*
> ```Shell
> #!/bin/bash
> 
> # Creates the ROSCATest.sol contract (with members publicized)
> ```

please update (my fault...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8)*
<!-- Sent from Reviewable.io -->

rmerom reviewed (approved):
Approving, let's merge this but add more tests in a subsequent PR (as you mentioned)

morelazers committed:
Fix most of the simpler PR review comments

morelazers committed:
Merge branch 'develop' into trustcoin_tests

morelazers commented:
Working on the rest but pushed the simpler fixes for now

---

Review status: 4 of 20 files reviewed at latest revision, 39 unresolved discussions.

---

*[contracts/ExampleTrustcoin2.sol, line 39 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLNxDimB867JxPGn6:-KfLKpaLTkTsH48BtmIU:bgrku7w) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/ExampleTrustcoin2.sol#L39)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

i'm not sure i understand what's wrong with having a constructor that takes the `_oldTokenAddress` , as long as `oldTokenAddress` is public and accessible?
</blockquote></details>

I think we talked about having it hardcoded so that people didn't have to "check" so to speak. If the old token given to the constructor is something malicious instead of the actual old token, we could migrate any old token to this contract.

---

*[contracts/Trustcoin.sol, line 26 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLZBsD46zTPex2Evo:-KfLLLsM4x0lsFl_OpEN:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L26)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

didn't we agree this can be removed?
</blockquote></details>

Done.

---

*[contracts/Trustcoin.sol, line 98 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLbOlj5H39Znf6zYP:-KfLLMiHCOGHHCgH9aaH:b-ie97g) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L98)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

why is this needed?
</blockquote></details>

Left in from testing, thought I had removed - thanks

---

*[contracts/Trustcoin.sol, line 112 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLguP4Y1gauWIhKbS:-KfLLok6GKDw3Rj_Ru57:b-a9txqq) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/Trustcoin.sol#L112)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

i think we should also assert that `newTokenAddress == 0`. Otherwise post-migration we can re-start migration.
If that's true, can you add that to the list of edgecase tests?
</blockquote></details>

I'll add that test. Assertion added for now.

---

*[contracts/deps/OutgoingMigrationTokenInterface.sol, line 15 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJM0cDtz3JGJe9iThI:-KfLMO_6nrpBi1aYWpGH:b-wwykf1) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/deps/OutgoingMigrationTokenInterface.sol#L15)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

these comments need refreshing... and this var isn't really used by the derived contract, so maybe we should rethink about how to do this?
</blockquote></details>

This is the one we're using in the above assertion, to ensure migrations can't be restarted after they're finalized (there's another way of doing it without this var though). It's also good to have this public so people can verify imo.

---

*[test/utils/utils.js, line 10 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJzOi893xAVt3taYvs:-KfLP1sR8BAoo01JAH7R:bi9sury) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/utils/utils.js#L10)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

s/becaues/because

(also I'm not sure i understand what this means)
</blockquote></details>

Removed (copied from rosca-contracts utils.js)

---

*[test/utils/utils.js, line 34 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJz_0AEd4aqrz80Ysa:-KfLPA8ptLjpLBmqz3HN:bqmul67) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/utils/utils.js#L34)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

unless we expect some exception to happen, can we rehtrow here?
otherwise we're swallowing exceptions
</blockquote></details>

Not using this anymore, removed.

---

*[test/checkInitialBalance.js, line 14 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK11NUWPW6al99zyT4:-KfLO0GIEiDm2yPp3vhh:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L14)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

"in the first account" - does that mean `msg.sender` ?
</blockquote></details>

Done.

---

*[test/checkInitialBalance.js, line 17 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK19FhYU3iS-6Po_DV:-KfLO5jgTq_pKERZQik_:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L17)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

I find that strings (here and below) that explain trivial things in the code just add more clutter. if a test fails, dev will go and find the line anyway (because they'll need to fix) and will likely understand from the code w/o needing the explanation.
I would start with having no string at all, and explain only when the test code is complicated enough for a reader not to understand what's going on.
</blockquote></details>

Done.

---

*[test/checkInitialBalance.js, line 27 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK1Pl-yf5GIPIahVYy:-KfLOIEgU7tEXivCwI1I:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L27)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

sorry, today I'm writing all these little things that you need to change if you don't want to :)

this is testing trivial code. If production code says `const a=6`, there's no point usually in testing that `a == 6`. It does add clutter, make maintainability harder (need to change in 2 places), and does not really check that a sophisticated piece of code is doing what it should. 
I think many of the following tests (though not all) might fall under this criterion.
</blockquote></details>

Done.

---

*[test/checkInitialBalance.js, line 39 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK27gWHpMGVYd-WpWf:-KfLOJX9sK6K15LfxxRZ:bss4a2b) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L39)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

... upon initialization?
</blockquote></details>

Removed.

---

*[test/checkSend.js, line 14 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4SjzbOZzRi2RpILs:-KfLOaDwYNW59XPJBJv6:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkSend.js#L14)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

many style guides discourage shortening words (e.g. "acc" for "account") because reader has to guess what the writer meant. If you don't want to make the var name too long, consider `account1BalanceStart`. same for other `acc` vars
</blockquote></details>

Done.

---

*[test/checkSend.js, line 20 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4uvn7kUwhhyIO5gO:-KfLOyZo1pw9GxhsZdd8:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkSend.js#L20)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

silly, but maybe also check `totalSupply` hasn't changed?
</blockquote></details>

Done.

---

*[test/checkAuth.js, line 1 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJzsmkjAr_96rwpq7k:-KfLN5SRRfcjIMLDq4f4:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L1)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

nit: since all tests start with `check`, can we drop it? all of them sit in the`test` directory so we know what they are about.
nit2: I would rather "auth" be expanded to "authentication" as i wasn't sure if it's "authentication" or "authorization" until reading the code (really nitty nit)
</blockquote></details>

Done.

---

*[test/checkAuth.js, line 18 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK-JgPmTpJnqGdf_Jh:-KfLNGpXMCPTSg_wkG7z:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L18)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

nit: would be nice to have 1 blank line between tests (here and in other test files), would also be consistent with rosca-contracts
</blockquote></details>

Done.

---

*[test/checkAuth.js, line 40 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK-lsmm_0Asc_QbmQf:-KfLNfQer_S49Cmt7Gsp:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkAuth.js#L40)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

nit: can use here and elsewhere `assert.isOk` for equal true, and `isNotOk` for equal false
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-:-KfLQOEiU_Eudq2uCmnE:b-zhi5mt)*
<!-- Sent from Reviewable.io -->

morelazers committed:
These files shouldn't be tracked afaik

morelazers commented:
Review status: 4 of 20 files reviewed at latest revision, 39 unresolved discussions, some commit checks failed.

---

*[test/authentication.js, line 47 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK0DkU17vqZwdYB8xg:-KfLVA0ba4sp73ynTAVh:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/authentication.js#L47)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

nit: you might want to query for `trst.minimumMigrationDuration` in a `before()` method and put it into a var `x`. Then use `x + <1 week>` when chcecking these things. Otherwise we have a constant here that really depends on a constant in the contract and may break if we decide to change the contract.
</blockquote></details>

Done.

---

*[test/migrationPeriod.js, line 15 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK4AC-9UqZafzmZLVo:-KfLXPxW_I7CimEC4719:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/migrationPeriod.js#L15)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

this can be made tighter - that is, wait `minMigrationPeriod - safetyDelta` and see that we still can't be migrate
</blockquote></details>

Done.

---

*[tools/build/run-tests.sh, line 3 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK501RuwLXBpklm4Ld:-KfLXbkRxgrVhOcqWcbV:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/tools/build/run-tests.sh#L3)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

please update (my fault...)
</blockquote></details>

Done.

---

*[test/checkInitialBalance.js, line 9 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK0lUcotwW65WC-IIe:-KfLXclb5j4tq6LOVnBH:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/checkInitialBalance.js#L9)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

same here as above, better use `trst.totalSupply.call()`. Future cloners of our contract will thank you (even if you won't know that), and also future readers of this test (who won't need to guess where this number came from).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8)*
<!-- Sent from Reviewable.io -->

morelazers committed:
Update tests to use less constants, and pull some meta out of the contract for ease of maintenance

morelazers committed:
Factor out more code

morelazers committed:
Give tokens to someone other than the owner/migration master before testing migrate functionality

morelazers committed:
Give everyone more starting ETH so I don't have to restart testrpc so often

morelazers committed:
End to end test for migration features

morelazers commented:
Review status: 3 of 28 files reviewed at latest revision, 39 unresolved discussions.

---

*[test/migrateTokens.js, line 66 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK3rKSzvLMwoUHx_B4:-KfLl-UW1A-3N-_3KDDp:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/migrateTokens.js#L66)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

I would add one longer test of a "normal scenario"
- migration begins
- people still transfer money in the old contract
- some people migrate, check `totalMigrated`and `totalSupply` on both
- migration is finalized
- no one can migrate, but can still `transfer` in both old and new contracts

perhaps in its own e2e test file?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8)*
<!-- Sent from Reviewable.io -->

morelazers committed:
Add test for restarting migration after finalisation

morelazers commented:
Review status: 3 of 28 files reviewed at latest revision, 39 unresolved discussions.

---

*[test/migrateTokens.js, line 37 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK37rnU176VujFFp5Y:-KfMCTaZG7bg0Jg0QtRj:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/migrateTokens.js#L37)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

this is very specific. IIUC 1) the migrator is the migrationMaster (can we do someone else?) and 2) their balance is zero (can we have balance `x != 0` and try to migrate `x+1`)?
</blockquote></details>

Done.

---

*[test/migrateTokens.js, line 48 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfK3ZCu0pw-yteUtqMU:-KfMCSDV43uPAbB7KGNA:b-896fix) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/test/migrateTokens.js#L48)):*
<details><summary><i>Previously, rmerom (Ron Merom) wrote…</i></summary><blockquote>

also pretty specific. can we transfer `x` to some account other than `owner`, then migrate `x` to new coin?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8)*
<!-- Sent from Reviewable.io -->

rmerom commented:
Review status: 3 of 28 files reviewed at latest revision, 38 unresolved discussions.

---

*[contracts/ExampleTrustcoin2.sol, line 39 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJLNxDimB867JxPGn6:-KfPbh7O1PzRnZe9hDgC:bssw3nc) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/ExampleTrustcoin2.sol#L39)):*
<details><summary><i>Previously, morelazers (Tom Nash) wrote…</i></summary><blockquote>

I think we talked about having it hardcoded so that people didn't have to "check" so to speak. If the old token given to the constructor is something malicious instead of the actual old token, we could migrate any old token to this contract.
</blockquote></details>

right, but i think people will have to check the deployed contract (reading a theoretical code would not help), and in case the `oldTokenAddress` is public, they can see it - even on [etherscan](https://etherscan.io/address/0xe0b7927c4af23765cb51314a0e0521a9645f0e2a#readContract) - no need to call the blockchain

---

*[contracts/deps/OutgoingMigrationTokenInterface.sol, line 15 at r3](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8#-KfJM0cDtz3JGJe9iThI:-KfPcfO7fbB92MfSYTzT:bq0ncok) ([raw file](https://github.com/wetrustplatform/token-contracts/blob/2c6aad4961c7db0db4da2c4a7363aed934aa2e90/contracts/deps/OutgoingMigrationTokenInterface.sol#L15)):*
<details><summary><i>Previously, morelazers (Tom Nash) wrote…</i></summary><blockquote>

This is the one we're using in the above assertion, to ensure migrations can't be restarted after they're finalized (there's another way of doing it without this var though). It's also good to have this public so people can verify imo.
</blockquote></details>

sounds good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wetrustplatform/token-contracts/8)*
<!-- Sent from Reviewable.io -->

rmerom reviewed (commented):
Thanks for doing all these fixes and adding so much code!
I actually meant that if you could fix the small comments and merge, I'd be happy to look at bigger pieces of code (like e2e test, and maybe other sources added?) in another PR. It's really good that we have as much code as possible reviewed.

rmerom commented:
btw, don't get me wrong: the new tests + the e2e all look great :)
